### PR TITLE
Support metallb (BGP/L2) in dev-scripts

### PR DESCRIPTION
**Pull request by s1061123**
This script provides to setup MetalLB/FRR to metal3 dev-scripts. 

This is mainly for [Service Load Balancers for On Premise Infrastructure enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/network/on-prem-service-load-balancers.md) and it is used not only dev/testing for MetalLB for OpenShift also OpenShift CI.

### CODE REVIEW

**Comment by russellb** [comment]
/ok-to-test

I can help review this

File: metallb/metallb.yaml.j2 [hunk 1]
```diff
@@ -0,0 +1,434 @@
+---
```
Comment (russellb) [inline metallb/metallb.yaml.j2:1]: What consumes this file?

We should also document somewhere that installation via manifest is a temporary measure and will be removed once an operator is available.  This is just to allow easier testing of metallb sooner in an openshift environment.

**Review by russellb — COMMENTED** [review]
Can you update the PR to refer to this enhancement for some context?  https://github.com/openshift/enhancements/blob/master/enhancements/network/on-prem-service-load-balancers.md

The description could also clarify the end goal of using dev-scripts as an environment to do metallb dev and testing for OpenShift, as well as a way to run metallb in OpenShift CI.

File: metallb/metallb.yaml.j2 (hunk 1 above)
Comment (s1061123) [reply metallb/metallb.yaml.j2:1]: This is for manifest for metalLB with parameterize image. `config_metallb_bgp.sh` and `config_metallb_l2.sh` has variables, METALLB_IMAGE_BASE and METALLB_IMAGE_TAG so we can easily specify which metallb container image is used.
 I believe that CI case we need to do so, that is why added.

I understand that manifest must be short life until metallb operator is introduced. Once we introduce metallb operator, of course we also need to change this script as well, so it is not a big problem for now.

Comment (s1061123) [reply metallb/metallb.yaml.j2:1]: I will update README to mention your comment (i.e. it is pretty temporally solution and metallb-operator will change this script)

Comment (s1061123) [reply metallb/metallb.yaml.j2:1]: Update the README.

**Comment by stbenjam** [comment]
I think it's reasonable to add an OWNERS under metallb so you can manage additional changes yourselves

**Comment by s1061123** [comment]
@stbenjam thanks, will do that!

**Comment by s1061123** [comment]
@russellb Update the PR diff because my shell scripts are ignored due to `.gitignore`. Sorry for confusing!

File: metallb/start_frr.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+sudo firewall-cmd --zone=libvirt --permanent --add-port=179/tcp
+sudo firewall-cmd --zone=libvirt --add-port=179/tcp
+
+# copy frr template
+cp -r ${metallb_dir}/frr_template ${metallb_dir}/frr
+
+# get node IP address
+# XXX: need to check v4/v6 dual stack case
+node_iplist=$(oc get node -o jsonpath='{.items[*].status.addresses[0].address}')
+python << EOF >> ${metallb_dir}/frr/bgpd.conf
```
Comment (markdgray) [inline metallb/start_frr.sh:31]: maybe you could add the generated files/folder to .gitignore?

File: metallb/config_l2.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+metadata:
+  namespace: metallb-system
+  name: config
+data:
+  config: |
+    address-pools:
+    - name: default
+      protocol: layer2
+      addresses:
+      - 192.168.10.0/24
```
Comment (markdgray) [inline metallb/config_l2.yaml:13]: Is this a standard metal3 address space?

File: metallb/start_frr.sh (hunk 2 above)
Comment (s1061123) [reply metallb/start_frr.sh:31]: changed.

**Comment by markdgray** [comment]
lgtm

**Comment by s1061123** [comment]
@markdgray thank you so much for reviewing this!

**Comment by russellb** [comment]
Can you add `metallb/OWNERS` ?

**Comment by s1061123** [comment]
@russellb Sure. will update PR to add me into dev-scripts/OWNERS. If I missed somewhere else, please let me know.

**Comment by russellb** [comment]
> @russellb Sure. will update PR to add me into dev-scripts/OWNERS. If I missed somewhere else, please let me know.

You can create a new `OWNERS` file in the `metallb` directory.  Go ahead and put my name in there too, in case I get removed from the top-level OWNERS at some point.  I haven't been helping with general dev-scripts in a long time.

**Comment by s1061123** [comment]
@russellb got it. I will create metallb/OWNERS and put my name.

File: metallb/OWNERS [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# See the OWNERS docs: https://git.k8s.io/community/contributors/guide/owners.md
+
+approvers:
+  - markmc
```
Comment (russellb) [inline metallb/OWNERS:4]: markmc might find himself surprised to be in here, though I'm not against it.  :-)

Maybe you meant markdgray ?

File: metallb/Makefile [hunk 5]
```diff
@@ -0,0 +1,21 @@
+.PHONY: config_metallb start_frr
+
+all:
+	echo "make {config_bgp | config_l2}"
+
+config_bgp: config_metallb_bgp start_frr
```
Comment (russellb) [inline metallb/Makefile:6]: Maybe `start_frr` should come before `config_metallb_bgp`?  To have the cleanest startup, we'd want frr running before deploying metallb.

You could make `start_frr` a dependency on `config_metallb_bgp`

```
config_bgp: start_frr
    ./metallb_bgp.sh

start_frr:
    ./start_frr.sh
```

File: metallb/Makefile [hunk 6]
```diff
@@ -0,0 +1,21 @@
+.PHONY: config_metallb start_frr
```
Comment (russellb) [inline metallb/Makefile:1]: need to add the rest of the targets in this Makefile to this list

File: metallb/OWNERS (hunk 4 above)
Comment (s1061123) [reply metallb/OWNERS:4]: Thank you for pointing it out! fixed.

File: metallb/Makefile (hunk 5 above)
Comment (s1061123) [reply metallb/Makefile:6]: Changed.

File: metallb/Makefile (hunk 6 above)
Comment (s1061123) [reply metallb/Makefile:1]: added.

**Comment by s1061123** [comment]
@russellb can we add @markdgray in metal3 group to add him into OWNER file?

**Comment by russellb** [comment]
> @russellb can we add @markdgray in metal3 group to add him into OWNER file?

added, though I think he's on PTO this week, so he won't be around to accept it until next week

**Comment by russellb** [comment]
I'm good with merging this now and iterating on it.  It won't affect any dev-scripts users or CI jobs unless they try to use it.

If you want to merge it now, you can remove mark gray from the OWNERS file and then submit that as a follow-up PR for us to merge once he's back from PTO to accept the github org invitation.

/approve

**Comment by s1061123** [comment]
I'm okey to wait a merge for his PTO back.

/retest

**Comment by s1061123** [comment]
@russellb now OWNER files are sync but need lgtm label. Can you give me lgtm?

**Comment by andfasano** [comment]
@s1061123 even if not strictly required for this PR, as discussed having a [configuration variable](https://github.com/openshift-metal3/dev-scripts/blob/master/config_example.sh) to enable the metal lb will greatly simplify the integration within the OpenShift CI, allowing to use the usual `DEVSCRIPT_CONFIG` CI variable to activate the required feature for a given job. That's the usual approach adopted across all the various OpenShift CI components that are defining tests based on dev-scripts, for example see this [job](https://github.com/openshift/release/blob/35920e882cb38dcae622c0a34d18380386046a3f/ci-operator/config/openshift/baremetal-operator/openshift-baremetal-operator-master.yaml#L69-L76)

**Comment by s1061123** [comment]
@andfasano I've added ENABLE_METALLB option.

File: 06_create_cluster.sh [hunk 7]
```diff
@@ -49,4 +49,22 @@ if [[ -n "${APPLY_EXTRA_WORKERS}" ]]; then
     fi
 fi
 
+if [[ "${ENABLE_METALLB}" == "true" ]]; then
+	if [[ ! -z "${$METALLB_MODE}" ]]; then
+		if [[ ${METALLB_MODE} == "l2" ]]; then
+			pushd metallb
+			make config_l2
```
Comment (andfasano) [inline 06_create_cluster.sh:56]: From the integration point of view with dev-scripts, this sounds a little bit unusual - ie using directly the related scripts seems also easier here

File: 06_create_cluster.sh [hunk 8]
```diff
@@ -49,4 +49,22 @@ if [[ -n "${APPLY_EXTRA_WORKERS}" ]]; then
     fi
 fi
 
+if [[ "${ENABLE_METALLB}" == "true" ]]; then
```
[the remark this answers is not in the document]
Comment (s1061123) [reply 06_create_cluster.sh]: changed.

File: 06_create_cluster.sh (hunk 7 above)
Comment (s1061123) [reply 06_create_cluster.sh]: changed.

File: 06_create_cluster.sh [hunk 9]
```diff
@@ -49,4 +49,19 @@ if [[ -n "${APPLY_EXTRA_WORKERS}" ]]; then
...
 
+if [[ ! -z "${ENABLE_METALLB_MODE}" ]]; then
+	if [[ ${ENABLE_METALLB_MODE} == "l2" ]]; then
+		pushd metallb
+		./metallb_l2.sh
+		popd
+	elif [[ ${ENABLE_METALLB_MODE} == "bgp" ]]; then
+		pushd metallb
+		./metallb_bgp.sh
+		./start_frr.sh
```
Comment (russellb) [inline 06_create_cluster.sh:60]: I would start frr first since the metallb speaker depends on it.

**Comment by russellb** [comment]
I'm fine doing further changes in follow-ups, since this is an optional off-by-default feature.

I'd like to see this migrated to using the metallb operator once the operator is ready.

/approve
/lgtm

**Comment by s1061123** [comment]
@russellb due to the changes of openshift/release (fix metallb CI), I also need to change scripts, so I'll incorporate your comment soon. Then I need your lgtm/approve again, so I'll let  you know.

**Comment by s1061123** [comment]
@russellb Use OPENSHIFT_RELEASE_IMAGE for metallb's container by default, for CI because now CI image contains metallb images by my fix PR https://github.com/openshift/release/pull/19958/files#diff-30a88cad2cb1057d80f437ce5f42b871f8e32d58ae942f44a65a90f444880f0dR15-R22

The additional code: 
https://github.com/openshift-metal3/dev-scripts/pull/1243/files#diff-330fd7ff1f29fa7dddcc0c0b2028826fa2f5596b333cb7b38e660b918ca96ed7R54-R57

**Comment by russellb** [comment]
> @russellb Use OPENSHIFT_RELEASE_IMAGE for metallb's container by default, for CI because now CI image contains metallb images by my fix PR https://github.com/openshift/release/pull/19958/files#diff-30a88cad2cb1057d80f437ce5f42b871f8e32d58ae942f44a65a90f444880f0dR15-R22
> 
> The additional code:
> https://github.com/openshift-metal3/dev-scripts/pull/1243/files#diff-330fd7ff1f29fa7dddcc0c0b2028826fa2f5596b333cb7b38e660b918ca96ed7R54-R57

Thanks!  This will break if you set dev-scripts to run against an older branch, but it's probably not worth the script complexity to try to validate which release image we're using.

/lgtm

**Pull request merged**